Set up backend

hajineats opened the pull request:
- Set up typescript node & express & mongodb
- Used npm for dependency management (so make sure to only use npm)
    - `npm i` then `npm start` to run the backend
- Used nested routing (express) to set up sample endpoints (refer to 750 wk3 modules for these)
- .env file should be placed under backend folder, and have MONGODB_URL, PORT
- Project structure:
    - src
        - models: has all mongodb database schema (eg person.model.ts)
        - routes: configures REST routes (eg person.route.ts)
        - controllers: defines async functions used by routes (eg person.controller.ts)
        - utils: configures environment values, middleware for logging and catching errors
        - index.ts: entrypoint that configures mongoose & express
- ESLint added for code style enforcing. Extended AirBnB style
    - Run lint through `npm run lint`, and fix lint issues through `npm run lint-fix`

hajineats committed:
chore: initial backend setup (#2)

butterplup commented on backend/.gitignore:
```diff
@@ -0,0 +1,134 @@
+node_modules
+.env
```
.env is already mentioned on line 80

butterplup replied:
and it seems node_modules is also already present on line 46

hajineats replied:
fair observation, thanks!

butterplup reviewed (changes requested):
Would you be able to add a newline character at the end of each file so that it meets the requirements for ESLint?

hajineats committed:
chore: clean up gitignore

hajineats committed:
refactor: leverage typescript and separate concerns

add types to parameters whenever possible
mongoose model to link to interface (for autosuggest)
add middleware for handling errors, and add logger
structure code into model, route, controller
remove require module imports

hajineats commented:
> Would you be able to add a newline character at the end of each file so that it meets the requirements for ESLint?

Yep done. It would probably be a good idea to leave any other formatting issues until eslint is set up for backend (#15)

hajineats committed:
Add Airbnb ESLint & fix formatting issues

butterplup reviewed (approved):
LGTM!

sydive reviewed (approved):
LGTM 👍 
🐜 We should probably change the folder name to `forgettable-backend` from `backend` at some point, but otherwise looks good!
